api/apps/api/src/modules/features/domain/feature-specification-revision.ts

```diff
@@ -0,0 +1,20 @@
+import {
+  ensure,
+  isDefined,
+  isInteger,
+  TinyType,
+  isGreaterThanOrEqualTo,
+} from 'tiny-types';
+
+export class FeatureSpecificationRevision extends TinyType {
+  constructor(public readonly value: number) {
+    super();
+    ensure(
+      'FeatureSpecificationRevision',
+      value,
+      isDefined(),
+      isInteger(),
+      isGreaterThanOrEqualTo(1),
+    );
+  }
+}
```

Comment:
though I like the idea behind tiny types, I would not do this while we are already using `class-validator` extensively or w open the door to using a slightly different hammer for each slightly different shape of nail. If we don't need to add validity guarantees a `type FirstName = string` should do, and if we want validation, I would use a class and decorate its properties via `class-validator`.

maybe in this specific case, defining an interface with a single `class-validator`-annotated revision property and implementing this in the FeatureSpecification class?

Reply:
`tiny-types` provide a bit more intention cases (as `type X = string` does not imply any safety when using the class, while `new Revision(3)` is a bit more expressive.

If we were to use `class-validator` it would imply more code than necessary - additional dto class and manual validation within the ValueObject (or even more, use single-propertied DTO ...)

The approach seems to be just more suitable for the needs - forcing ourselves to use library we use somewhere else isn't always the best way. Tools should help us rather than dictate how we implement things, especially when it comes to such small "cases" - we could paraphrase hammer metaphor - "why to use same heavy hammer, while just finger is needed" ? :)

Reply:
yes, indeed I am on the fence on this one. I would even favour a bit more verbosity/more moving pieces over more ways to do something (or something similar, I agree, the use case is different), for the sake of ongoing maintainability.

If we limit this to the indended use case, that is individual primitive types then it makes sense. Time to start `CODING_STYLE.md`, I guess 🤔

Reply:
Actually, it's possible to use both libs without extra DTO class:
```
const ClassValidatorTinyTypeOf = <T>() => {
  return class extends TinyTypeOf<T>() {
    constructor(value: T) {
      super(value);
      const result = validateSync(this);
      if (result.length > 0) {
        throw result;
      }
    }
  };
};

class ATinyType extends ClassValidatorTinyTypeOf<number>() {
  @IsDivisibleBy(1)
  value!: number;
}
```

But it is artificial a bit, and I agree with @kgajowy that using the validator from tiny types is a better fit here. I'd keep class-validator to define validators for DTOs as it plays well with built-in nestjs behavior and tiny-types for domain objects. I think we need to find a balance to not use class-validator as Maslow's hammer and to not mix too many libs for validating unexpectedly.

The last important thing is that tiny-types are not MIT, but Apache 2.0. Is it a problem @hotzevzl?

Reply:
@Dyostiq your example is pretty much the first thing that came to my mind when looking at the `tiny-types` documentation, but yes, the extra verbosity makes the "tiny" selling point less tiny.

Let's apply the rule of three to these contexts (vs code duplication) - if we get to a third way of doing validation then we should all get a flashing red alert lighting up.

Reply:
as for the package's license, from my understanding it is ok - do you have specific concerns?